### PR TITLE
Zuul http method routing. Issue #1331

### PR DESCRIPTION
**Pull request by galovics**
First implementation for this enhancement. It's not yet finished as I want to add more tests to cover the changes but I wanted to have a quick feedback whether this way of implementation is acceptable for you guys.

### CODE REVIEW

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/RequestWrapper.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		return result;
+	}
+
+	@Override
+	public String toString() {
+		return "RequestWrapper{" + "path='" + path + '\'' + ", method=" + method + '}';
+	}
+
+	public static RequestWrapper fromPath(String path) {
+	    Objects.requireNonNull(path, "path must not be null");
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/RequestWrapper.java]: spacing

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java [hunk 2]
```diff
@@ -16,22 +16,28 @@
...
-import java.util.Collection;
-import java.util.LinkedHashMap;
-import java.util.List;
-import java.util.Map;
+import static java.lang.String.format;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.Collectors.toSet;
+
+import java.util.*;
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: no `*` imports

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java [hunk 3]
```diff
@@ -109,26 +118,34 @@ protected Route getSimpleMatchingRoute(final String path) {
...
+					Map<HttpMethod, ZuulRoute> methods = entry.getValue();
+					return methods.entrySet().stream().filter(e -> matchesMethod(method, e)).findAny().map(Entry::getValue).orElse(null);
 				}
 			}
 		}
 		return null;
 	}
 
+	private boolean matchesMethod(HttpMethod method, Entry<HttpMethod, ZuulRoute> zuulRoute) {
+		if (method == null) {
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: is this even possible?

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/ZuulProperties.java [hunk 4]
```diff
@@ -16,24 +16,20 @@
 
 package org.springframework.cloud.netflix.zuul.filters;
 
-import com.netflix.hystrix.HystrixCommandProperties.ExecutionIsolationStrategy;
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/ZuulProperties.java]: no `*` imports and make sure import order is correct

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/discovery/DiscoveryClientRouteLocator.java [hunk 5]
```diff
@@ -16,9 +16,10 @@
 
 package org.springframework.cloud.netflix.zuul.filters.discovery;
 
-import java.util.LinkedHashMap;
-import java.util.List;
-import java.util.Map;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+
+import java.util.*;
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/discovery/DiscoveryClientRouteLocator.java]: no `*` imports

File: spring-cloud-netflix-zuul/src/test/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocatorTests.java [hunk 6]
```diff
@@ -16,97 +16,101 @@
 
 package org.springframework.cloud.netflix.zuul.filters;
 
-import java.util.ArrayList;
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-zuul/src/test/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocatorTests.java]: import order

**Comment by galovics** [comment]
Thanks for taking a look.

1. Sure, you mean standard docs or javadoc as well?
2. So basically now for each route we are having a list of methods which are matching. In case you didn't specify the method, we will fill up the list with all the methods available in the HttpMethod enum.

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java (hunk 3 above)
Comment (galovics) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: It is, as now the RouteLocator interface defines 2 getMatchingRoute methods. The first one is the original one which I made deprecated and the new one is a *default* method with the following implementation:
`return getMatchingRoute(request.getPath());`
meaning that we won't have method here.

This means that if someone already had a special RouteLocator implementation, we don't want to break that but just falling back to it's original implementation.

The reason is that I just didn't want to introduce any breaking API changes, but if you are OK with that, I can remove the original method from the interface.

Comment (ryanjbaxter) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: This is going into the 2.0.x release so this is the time to make breaking API changes

**Comment by ryanjbaxter** [comment]
I mean the docs in the docs folder. 

OK so that means any method request to that route will match.  great.

**Comment by galovics** [comment]
Docs added yesterday.

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java (hunk 3 above)
Comment (galovics) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: Okay, then I'll introduce the breaking change.

However, does it make sense to go with this implementation for 1.x (next release)?

When do you guys plan to release 2.0?

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java [hunk 7]
```diff
@@ -16,26 +16,38 @@
...
 import org.springframework.cloud.netflix.zuul.filters.ZuulProperties.ZuulRoute;
 import org.springframework.cloud.netflix.zuul.util.RequestUtils;
 import org.springframework.core.Ordered;
+import org.springframework.http.HttpMethod;
 import org.springframework.util.AntPathMatcher;
+import org.springframework.util.CollectionUtils;
 import org.springframework.util.PathMatcher;
 import org.springframework.util.StringUtils;
 
+import com.google.common.collect.Sets;
```
Comment (spencergibb) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: Please remove `com.google` references

Comment (galovics) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: Okay, didn't know that's forbidden to use. Does it make sense to remove Guava from the dependencies?

**Comment by galovics** [comment]
@spencergibb what's your point? Should I separate the new functionality somehow into a separate class?

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java (hunk 7 above)
Comment (spencergibb) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: That is what I'm asking

Comment (galovics) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: Maybe I wasn't clear enough with 'dependencies'. I meant the Maven dependencies. Of course I'll remove it from the imports but maybe it makes sense to remove the Maven dep as well.

Comment (spencergibb) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: Yes

Comment (galovics) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java]: Just checked it and it seems that the Guava lib is coming through this chain:
    [INFO] +- com.netflix.zuul:zuul-core:jar:1.3.1:compile
    [INFO]  |  +- com.netflix.archaius:archaius-core:jar:0.7.5:compile
    [INFO]  |   |  +- com.google.guava:guava:jar:18.0:compile

So I'll leave it as it is and remove the import only.

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/PreDecorationFilter.java [hunk 8]
```diff
@@ -23,13 +23,11 @@
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
-import org.springframework.cloud.netflix.zuul.filters.ProxyRequestHelper;
-import org.springframework.cloud.netflix.zuul.filters.Route;
-import org.springframework.cloud.netflix.zuul.filters.RouteLocator;
-import org.springframework.cloud.netflix.zuul.filters.ZuulProperties;
+import org.springframework.cloud.netflix.zuul.filters.*;
```
Comment (spencergibb) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/PreDecorationFilter.java]: No `*` imports please

**Comment by spencergibb** [comment]
See https://github.com/spring-cloud/spring-cloud-gateway/tree/master/spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/handler/predicate for how I used `java.util.Predicate`

**Comment by galovics** [comment]
Applied the requested changes. 
@spencergibb I agree with you, however I think there should be a different technical issue to refactor this part of the codebase to separate the behavior.

**Comment by spencergibb** [comment]
@galovics why? Why merge just to refactor?

**Comment by galovics** [comment]
Personally I think I'm not the best person for refactoring this whole part of the codebase as I just met it for the first time and it would take more time for me than for another person who has experience with this code.

Do you think its impossible to merge the PR in the current state?

**Comment by galovics** [comment]
Can I have some feedback about merging this PR without refactoring the whole routing logic? Thanks!

**Comment by spencergibb** [comment]
Marked for team discussion

**Comment by galovics** [comment]
Any updates maybe?

**Comment by ryanjbaxter** [comment]
@galovics we didnt get to it this week in our team meeting, I have added it to our agenda for next week

**Comment by spencergibb** [comment]
@galovics with the upcoming release of spring-cloud-gateway we don't want to be adding and supporting significant new features to the zuul 1 support.

**Comment by galovics** [comment]
@spencergibb thank you for the response from you guys. I think in this case it's wise to close any existing open issues which are referring to new features to avoid this situation again.

Btw just a quick question, is this a big change or why did you choose to let this PR go away? I think it's not a big change with low risk but Zuul was always lacking this feature.

**Pull request closed without merging**